src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+            for (int index = 0; index < warning1.length; index++) {
+                if (warning2[index].getPreferredPosition() != warning1[index].getPreferredPosition())
+                    return false;
+            }
+            return true;
+        };
+
+        // Stack traces are ordered top to bottom, and so duplicates always have the same first element(s).
+        // Sort the stack traces lexicographically, so that duplicates immediately follow what they duplicate.
+        final Comparator<DiagnosticPosition[]> ordering = (warning1, warning2) -> {
+            for (int index1 = 0, index2 = 0; true; index1++, index2++) {
+                final boolean end1 = index1 >= warning1.length;
+                final boolean end2 = index2 >= warning2.length;
+                if (end1 && end2)
+                    return 0;
+                if (end1)
+                    return -1;
+                if (end2)
+                    return 1;
+                final int posn1 = warning1[index1].getPreferredPosition();
+                final int posn2 = warning2[index2].getPreferredPosition();
+                final int diff = Integer.compare(posn1, posn2);
+                if (diff != 0)
+                    return diff;
+            }
+        };
+        this.warningList.sort(ordering);
+
+        // Now emit the warnings, but skipping over duplicates as we go through the list
+        DiagnosticPosition[] previous = null;
+        for (DiagnosticPosition[] warning : this.warningList) {
+
+            // Skip duplicates
+            if (previous != null && extendsAsPrefix.test(previous, warning))
+                continue;
+            previous = warning;
+
+            // Emit warnings showing the entire stack trace
+            JCDiagnostic.Warning key = Warnings.PossibleThisEscape;
+            int remain = warning.length;
+            do {
+                final DiagnosticPosition pos = warning[--remain];
+                this.log.warning(Lint.LintCategory.THIS_ESCAPE, pos, key);
+                key = Warnings.PossibleThisEscapeLocation;
+            } while (remain > 0);
+        }
+        this.warningList.clear();
+    }
+
+    private void analyzeStatements(List<JCStatement> stats) {
+        for (JCStatement stat : stats) {
+
+            // Analyze statement
+            this.scan(stat);
+
+            // Capture any pending warning generated
+            if (this.copyPendingWarning())
+                break;                      // report at most one warning per constructor
+        }
+    }
+
+    @Override
+    public void scan(JCTree tree) {
+
+        // Check node
+        if (tree == null || tree.type == Type.stuckType)
+            return;
+
+        // Sanity check
+        Assert.check(this.checkInvariants(true, false));
+
+        // Can this expression node possibly leave a 'this' reference on the stack?
+        final boolean referenceExpressionNode;
+        switch (tree.getTag()) {
+        case CASE:
```

Comment:
I put it there because of switch expressions and `yeild`... ?

Reply:
Well, yield can... yield a value - `case` doesn't. So I'm confused. Also because the variable is called `referenceExpressionNode` and `CASE` is not an expression. Can `CASE` leave anything on the stack? YIELD does, but CASE?

Reply:
It's just an artifact of the way switch expressions are handled:
```java
    @Override
    public void visitSwitchExpression(JCSwitchExpression tree) {
        visitScoped(tree, true, t -> {
            scan(t.selector);
            refs.discardExprs(depth);
            RefSet<ExprRef> combinedRefs = new RefSet<>();
            for (List<JCCase> cases = t.cases; cases.nonEmpty(); cases = cases.tail) {
                scan(cases.head);
                combinedRefs.addAll(refs.removeExprs(depth));
            }
            refs.addAll(combinedRefs);
        });
    }

    @Override
    public void visitCase(JCCase tree) {
        scan(tree.stats);          // no need to scan labels
    }
```
After scanning a switch expression case, the `yield`'ed value will be on the top of the stack.

Then `visitCase` does NOT remove it, so that it can be picked up and removed by `visitSwitchExpression()`.

But this can be cleaned up a little bit, by having `visitSwitchExpression()` not delegate to `visitCase()` but rather iterate through the case statements itself directly. Then we can remove the `CASE` as you suggest.

Also I realized we need to handle `yield`'ed values like `return` values, i.e., collect and remember them until the entire case is complete.

I'll fix.